### PR TITLE
Use python3 venv for installation.

### PR DESCRIPTION
**Pull request by ChrisCarini**
Suggest to use python3 venv for installation. This allows users to have non-conflicting dependencies in their base python installation.

**Note:** I am using `.venv` directory, as that is already excluded in the `.gitignore` file.

### CODE REVIEW

**Comment by ChrisCarini** [comment]
Hi @lorenzodifuccia - when you have a moment, would you mind considering this PR? Much appreciated!

**Comment by holzkohlengrill** [comment]
I think that's not what the standard user wants (also I don't have any conflicts).
As in general for python everybody is free to install it into a venv (which you need to activate before every use which is not that user friendly as the standard behaviour).

To sum it up, in my opinion, there should be the option to do so but this should be no obligation for a venv.

**Comment by ChrisCarini** [comment]
Hi @holzkohlengrill - thank you for sharing your thoughts!

> also I don't have any conflicts

This may be the case for you and folks, but is not necessarily going to be true for everyone. Having users install packages outside of a virtual environment can become problematic with people start using / working on many python projects where version conflicts may occur when doing this. For example, say I was working on another project that had a `requirements.txt` file set to `requests==1.2.3` (yes, I realize this is a very old version) and installed without a virtual environment, and then did similar for this project - my prior project would likely have issues. For a 'standard user', this may cause difficulties debugging, and may eventually result in a GitHub issue opened for whatever project they are encountering the issue (ie, it might not be https://github.com/lorenzodifuccia/safaribooks where the issue is opened).


I agree with your general sentiment, having a virtual environment is extra steps for the standard user and a bit more cumbersome for them. If we were to abstract these extra steps away for users (ie, if we created a nicer 'installation script', and wrapper script around the main python files that automatically sourced the venv prior to executing), would you still have concerns with using virtual environments?

**Comment by holzkohlengrill** [comment]
Thanks for the extensive response.

First I would say when a project uses old versions it's their duty to make it compatible to newer versions of the libraries.
Secondly having a separate install script for venv is additional maintaing effort with little benefit (however at the end @lorenzodifuccia has to decide this).
Third, I agree with the hassle about version conflicts. In the other hand you should avoid that in the first place (supporting current versions) instead of workarounds. Users having a multitude of scripts (installed) normally are capable of setting up a venv (also a quick Google search will show the steps for those who don't).

So for a big projects having complex setup routines this makes definitely sense.
@lorenzodifuccia what do you think?

**Comment by lorenzodifuccia** [comment]
Thank you so much @ChrisCarini @holzkohlengrill for your considerations!
I just think that a `venv` would add only an additional step for normal user, as said by Marcel.
Of course, people is free to manage their python libraries however they want.

Versions on `requirements.txt` will always be `>=` and with the first working version (addressing bugs or security issues).

We can add a line in the README encouraging people to use pipenv or a separate venv, in the Req&Setup cap...

**Comment by ChrisCarini** [comment]
_**TL;DR:** SGTM - I'll close and delete this branch then. Sorry to consume the time._

> Thanks for the extensive response.

@holzkohlengrill  - no problem!

> First I would say when a project uses old versions it's their duty to make it compatible to newer versions of the libraries.

I agree, in the ideal world this is how it would work. But as I'm sure you are aware, there are many still-useful projects that do not get the attention they deserve. This can be true in the OSS community, but also behind corporate walls, as you may or may not be aware.

> Secondly having a separate install script for venv is additional maintaing effort with little benefit (however at the end @lorenzodifuccia has to decide this).

Sure, understood that this would be additional effort - the benefit, IMO, would be keeping this project as isolated from other python projects on a users system as possible.

> Third, I agree with the hassle about version conflicts. In the other hand you should avoid that in the first place (supporting current versions) instead of workarounds. Users having a multitude of scripts (installed) normally are capable of setting up a venv (also a quick Google search will show the steps for those who don't).

Again, agree on avoiding, but this is not always possible as mentioned above.


> Thank you so much @ChrisCarini @holzkohlengrill for your considerations!

@lorenzodifuccia - Of course!

> I just think that a venv would add only an additional step for normal user, as said by Marcel.

SGTM - I'll close and delete this branch then. Sorry to consume the time.

> Versions on requirements.txt will always be >= and with the first working version (addressing bugs or security issues).

Understand - however my concern was not this project, per say, using `==` in the req.txt file, but rather other projects setting hard dependencies.


@lorenzodifuccia / @holzkohlengrill - what are both of your thoughts on making this a python zipapp (something like https://github.com/pantsbuild/pex or https://github.com/linkedin/shiv as examples), so the need for people to do the installation of dependencies (when using this repo; not when developing) becomes moot? Or is the intent/desire to keep this as a 'standalone' script?
